### ISSUE TITLE
''ffmpeg version git-... the FFmpeg developers built with gcc 9.3.1 (GCC) 20200523' Exception thrown when piping from and to stream

### ISSUE THREAD

**Issue by xinyuewu** [rosenbjerg/FFMpegCore#70]
**Version : 2.0.1**, installed via nuget

I'm getting this exception message when trying to convert videos piping from and out to MemoryStream. Please see as below:

```
 public byte[] ConvertVideo(Stream inputStream)
 {
     var arguments = FFMpegArguments
     .FromPipe(new StreamPipeSource(inputStream))
     .WithVideoCodec(VideoCodec.LibX264)
     .Scale(VideoSize.Ld);

     using var outputStream = new MemoryStream();
     var streamPipeDataReader = new StreamPipeSink(outputStream);
     var processor = arguments.OutputToPipe(streamPipeDataReader);
     var success = processor.ProcessSynchronously();

     return outputStream.ToArray();
}
```
The exception message is 

> "FFMpegCore.Exceptions.FFMpegException: 'ffmpeg version git-2020-06-03-b6d7c4c Copyright (c) 2000-2020 the FFmpeg developers
>   built with gcc 9.3.1 (GCC) 20200523
>   configuration: --enable-gpl --enable-version3 --enable-sdl2 --enable-fontconfig --enable-gnutls --enable-iconv --enable-libass --enable-libdav1d --enable-libbluray --enable-libfreetype --enable-libmp3lame --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libopus --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libsrt --enable-libtheora --enable-libtwolame --enable-libvpx --enable-libwavpack --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libzimg --enable-lzma --enable-zlib --enable-gmp --enable-libvidstab --enable-libvmaf --enable-libvorbis --enable-libvo-amrwbenc --enable-libmysofa --enable-libspeex --enable-libxvid --enable-libaom --disable-w32threads --enable-libmfx --enable-ffnvcodec --enable-cuda-llvm --enable-cuvid --enable-d3d11va --enable-nvenc --enable-nvdec --enable-dxva2 --enable-avisynth --enable-libopenmpt --enable-amf
>   libavutil      56. 49.100 / 56. 49.100
>   libavcodec     58. 90.100 / 58. 90.100
>   libavformat    58. 44.100 / 58. 44.100
>   libavdevice    58.  9.103 / 58.  9.103
>   libavfilter     7. 84.100 /  7. 84.100
>   libswscale      5.  6.101 /  5.  6.101
>   libswresample   3.  6.100 /  3.  6.100
>   libpostproc    55.  6.100 / 55.  6.100
> \\.\pipe\FFMpegCore_cae723fa-0c8c-4ec8-8ede-9668006b2432: Invalid data found when processing input'"
> 

How can I fix it? Your help is greatly appreciated!

Regards,
Claire

**Comment by rosenbjerg** [comment]
Have you tried using a stable version of ffmpeg? I can see that you are a git build, `git-2020-06-03-b6d7c4c`. I don't that is the cause, but I would be nice to be sure.

We have a unit-test that is somewhat similar to your use-case: [Video_TranscodeInMemory](https://github.com/rosenbjerg/FFMpegCore/blob/master/FFMpegCore.Test/VideoTest.cs#L622), though it uses a `RawVideoPipeSource` instead of `StreamPipeSource`.
What kind of stream are you inputting from?

**Comment by xinyuewu** [comment]
Hi there, thanks for you quick reply!

I've just tried with the latest release version 4.2.3 static builds (downloaded from https://ffmpeg.zeranoe.com/builds/), and the exception message are mostly the same except the first line "FFMpegCore.Exceptions.FFMpegException: 'ffmpeg version 4.2.3 Copyright ...."

I'm building a web API using netcore3.1. In the post request, an uploaded file is accepted as IFormFile and then transformed as MemoryStream.

I cloned your repository and all the unit tests in VideoTest ran successfully. So it's unlikely the error is caused by ffmpeg. 

I've also tried a few different cases:
1. Using FromInputFiles and OutputToFile with my local video, the function worked as expected;
2. Using FromPipe and OutputToFile with the same video uploaded through postman, the exception message became "Output file was not created"; 
3. Using FromInputFiles and OutputToPipe, the exception message became one as below:

> FFMpegCore.Exceptions.FFMpegException: 'ffmpeg version 4.2.3 Copyright (c) 2000-2020 the FFmpeg developers
  built with gcc 9.3.1 (GCC) 20200523
  configuration: --enable-gpl --enable-version3 --enable-sdl2 --enable-fontconfig --enable-gnutls --enable-iconv --enable-libass --enable-libdav1d --enable-libbluray --enable-libfreetype --enable-libmp3lame --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libopus --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libtheora --enable-libtwolame --enable-libvpx --enable-libwavpack --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libzimg --enable-lzma --enable-zlib --enable-gmp --enable-libvidstab --enable-libvorbis --enable-libvo-amrwbenc --enable-libmysofa --enable-libspeex --enable-libxvid --enable-libaom --enable-libmfx --enable-amf --enable-ffnvcodec --enable-cuvid --enable-d3d11va --enable-nvenc --enable-nvdec --enable-dxva2 --enable-avisynth --enable-libopenmpt
  libavutil      56. 31.100 / 56. 31.100
  libavcodec     58. 54.100 / 58. 54.100
  libavformat    58. 29.100 / 58. 29.100
  libavdevice    58.  8.100 / 58.  8.100
  libavfilter     7. 57.100 /  7. 57.100
  libswscale      5.  5.100 /  5.  5.100
  libswresample   3.  5.100 /  3.  5.100
  libpostproc    55.  5.100 / 55.  5.100
Input #0, mov,mp4,m4a,3gp,3g2,mj2, from 'C:\XinyueWu\eCom Scotland\Test files\Videos\input.mp4':
  Metadata:
    major_brand     : isom
    minor_version   : 512
    compatible_brands: isomiso2avc1mp41
    creation_time   : 1970-01-01T00:00:00.000000Z
    encoder         : Lavf53.24.2
  Duration: 00:00:13.50, start: 0.000000, bitrate: 1248 kb/s
    Stream #0:0(und): Video: h264 (Main) (avc1 / 0x31637661), yuv420p, 1280x720 [SAR 1:1 DAR 16:9], 862 kb/s, 25 fps, 25 tbr, 12800 tbn, 50 tbc (default)
    Metadata:
      creation_time   : 1970-01-01T00:00:00.000000Z
      handler_name    : VideoHandler
    Stream #0:1(und): Audio: aac (LC) (mp4a / 0x6134706D), 48000 Hz, 5.1, fltp, 381 kb/s (default)
    Metadata:
      creation_time   : 1970-01-01T00:00:00.000000Z
      handler_name    : SoundHandler
[NULL @ 00000198ea52dac0] Unable to find a suitable output format for '\\.\pipe\FFMpegCore_e2e403d5-4a84-44ab-b78e-3c849dbfcf3c'
\\.\pipe\FFMpegCore_e2e403d5-4a84-44ab-b78e-3c849dbfcf3c: Invalid argument'

Any idea what's causing the problem?

Thanks,
Claire

**Comment by rosenbjerg** [comment]
Hi again, 
The error logs are different this time: `[NULL @ 00000198ea52dac0] Unable to find a suitable output format for` (3rd last line). That error might indicate that ffmpeg is unable to determine the format, and needs to be told what format to use.
Could you try adding a call to `.ForceFormat(VideoType.Mp4)` or `.ForceFormat(VideoType.WebM)` to you argument builder?

Another thing worth considering:
FFMpeg might not be able to process the input stream as fast as the uploading client can upload, so it may bottleneck the upload.
For similar use-cases, I've first saved the incoming stream (from the IFormFile) to a local file, then processed it with ffmpeg, and then deleted the local file. This way, ffmpeg won't slow down the upload (nor can the upload slow down ffmpeg).

**Comment by rosenbjerg** [comment]
Any progress?

**Comment by rosenbjerg** [comment]
@xinyuewu Did you manage to solve the problem? :)

**Comment by xinyuewu** [comment]
Hi there,

I'm sorry for the late reply, and thank you for much the suggestions. 

We managed to bypass the problem by storing the input stream in a local temp file and process it with ffmpeg in the background, just like you have suggested and it is working awesome! Thanks!

I saw that "Convert to and from streams" is now added to README.md, so I guess the issue will probably be fixed for us if it is upgraded, but sorry I've not got time to try it yet. 

Anyways thank you very much for the great help and support!
Claire
